### PR TITLE
Search for repos by topic

### PR DESCRIPTION
**Pull request by spier**
Implements #11.

This assumes that `repos.json` will contain a property like this for every repo:
```
"topics": [ "some-topic", "and-another-topic", "and-even", "more", "topics" ]
```

## Changes:

* Support search by topic
* Show topics in modal window together with other repo details.

## Todo:

- [ ] describe in the documentation that the `topics` property needs to be available for every repo
- [ ] implement condition check to only display topics when they are there

### CODE REVIEW

**Comment by spier** [comment]
@mehrdadrad Take a look at this when you get a chance, to check if this is implementing this feature as you imagined it to work. 

You should be able to try it out based on [this branch](https://github.com/spier/project-portal-for-innersource/tree/issues/11-search-topics). When collecting the data for your `repos.json`, make sure to include the `topics` property for every repo, potentially using the `application/vnd.github.mercy-preview+json` media type when making GitHub API requests (see [docs here](https://docs.github.com/en/rest/reference/repos#get-all-repository-topics)).

**Comment by mehrdadrad** [comment]
Excellent, thanks for quick action! 
LGTM

**Comment by spier** [comment]
@Michadelic my implementation is currently just the happy path, without any consideration for edge cases. Also lacks documentation.

Would you mind reviewing this and letting me know what I should add to get this PR mergeable?

File: index.html [hunk 1]
```diff
@@ -197,6 +197,7 @@
             <h4>[[title]]</h4>
             <p class="repoLink"><a href="[[repoURL]]" target="_blank">[[repoTitle]]</a></p>
             <p>[[description]]</p>
+            <p><b>Topics:</b><br/>[[topics]]</p>
```
Comment (Michadelic) [inline index.html:200]: might be good to visualize this with a tag-like layout, e.g. a div with a certain class. Here is a suggestion:

``` html
<span class="pill">Topic</span>
```
``` css
.pill {
    border-radius: 2em;
    padding: 2px 10px;
    background-color: lightblue;
}
```

**Review by Michadelic — APPROVED** [review]
see layout suggestions, otherwise great!

File: index.html (hunk 1 above)
Comment (Michadelic) [inline index.html:200]: add `[[#if topics]] ... [[/if]]` so that this section is hidden when no topics are collected

**Review by Michadelic — APPROVED** [review]
just minor layout suggestions

File: index.html (hunk 1 above)
Comment (Michadelic) [reply index.html:200]: topic pills

Comment (spier) [reply index.html]: Nice! The styling looks a bit funky when the repo has a lot of topics but it is certainly an improvement already!

Screen Shot 2021-01-25 at 22 14 06

Comment (spier) [reply index.html]: I worked that in.
A bit unsure if I get that IF-replacement logic right but it seems to work.
See: https://github.com/SAP/project-portal-for-innersource/pull/12/files#diff-7a094af86da034414240a174bdfb685fe3cf7c50ececa1b23d4b3e91523554edR286-R288

**Comment by spier** [comment]
@Michadelic something that I discovered while developing on this:
I accidentally committed (and then reverted) a different local version of `repos.json`. I tried to add repos.json to `.gitignore` but that didn't work, maybe because the file is already under version control? A bit cumbersome, at least when testing with different datasets locally.

File: index.html (hunk 1 above)
Comment (Michadelic) [reply index.html]: it may still need some margin-bottom (5px or so should be fine) :-)

**Comment by Michadelic** [comment]
> @Michadelic something that I discovered while developing on this:
> I accidentally committed (and then reverted) a different local version of `repos.json`. I tried to add repos.json to `.gitignore` but that didn't work, maybe because the file is already under version control? A bit cumbersome, at least when testing with different datasets locally.

I see an entry for `repos.json` in `.gitignore`:
https://github.com/SAP/project-portal-for-innersource/blob/main/.gitignore#L2

Is it in the wrong notation. At least in my IDE it works as designed

File: scripts/index.js [hunk 2]
```diff
@@ -283,6 +283,10 @@ function showModal (vRepoId, oEvent) {
           : "";
   sHTML = sHTML.replace("[[description]]", sDescription);
 
+  let sTopics = oRepo.topics ? oRepo.topics.map((t) => "<span class='pill'>"+t+"</span>").join(" ") : "";
+  sHTML = sHTML.replace("[[topics]]", sTopics);
+  sHTML = sHTML.replace(/\[\[#if topics\]\](.*)\[\[\/if\]\]/, sTopics ? "$1" : "");
```
Comment (Michadelic) [inline scripts/index.js:288]: yes, that should work. It is a quick & dirty templating solution

File: styles/global.css [hunk 3]
```diff
@@ -558,3 +558,10 @@ tr:hover {
 .modal-overlay {
   background: #0a3050;
 }
+
+/* little blue pill around topics in details view */
+.pill {
+    border-radius: 2em;
+    padding: 2px 10px;
+    background-color: lightblue;
```
Comment (Michadelic) [inline styles/global.css:566]: add a margin-bottom of 5px or so to have som spacing between the lines

**Review by Michadelic — APPROVED** [review]
almost perfect :-)

File: styles/global.css (hunk 3 above)
Comment (spier) [reply styles/global.css:566]: Needed to add some more CSS, but now it works.

File: scripts/index.js (hunk 2 above)
Comment (spier) [reply scripts/index.js:288]: Yeah, I figured :)

**Comment by spier** [comment]
As a quick way to generate topics for each repo, I ended up copying properties that the repo already has into the `topics` property.

How it did it:
`cat repos.json | jq '[.[] | . + { "topics" : [.name, .language, .owner.login] }]'`

This results in e.g. "Earth" having these topics now:
```
"topics": [
   "earth",
   "JavaScript",
   "Sol"
]
```

I also modified the planets/repos Mars, Ceres, Venus manually, in order to test the edge cases of
 
- many topics
- no topics (empty topics array)
- no topics property

If you run this branch locally now, you can see it in full effect.

Example:

Screen Shot 2021-01-26 at 15 35 28

@Michadelic if you are happy with how this works now, feel free to merge.

**Review by Michadelic — COMMENTED** [review]
is there also one without the property `topics` (test case for not implementing this feature)
Looks goot otherwise, i would maybe limit the tags to 5 or so for the demo data

**Comment by spier** [comment]
Yes, that case is covered as well.

I used the first 4 planets/repos that are shown at the very top when launching the app.

* **Earth** - few topics (only 3) - happy path
* **Mars** - many topics (13)
* **Venus** - no topics (empty topics array)
* **Ceres** - no topics property

All repos have exactly 3 topics (with exception of the ones listed above).
**Mars** has that many repos to test how the line-breaks and formatting works with that many topics, so I would suggest to leave it that way.

**Review by Michadelic — APPROVED** [review]
👍
